### PR TITLE
Add HTTPS to REST API

### PR DESCRIPTION
**Pull request by charlesbvll**
#### What does this implement/fix? Explain your changes.

Gives the possibility to start the fleet API server with an SSL certificate in order to connect via HTTPS.

### CODE REVIEW

File: src/py/flwr/server/app.py [hunk 1]
```diff
@@ -556,3 +565,13 @@ def _add_args_fleet_api(parser: argparse.ArgumentParser) -> None:
         help=f"Fleet API REST server address. Default:'{ADDRESS_FLEET_API_REST}'",
         default=ADDRESS_FLEET_API_REST,
     )
+    rest_group.add_argument(
+        "--ssl_certfile",
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:569]:
```suggestion
        "--ssl-certfile",
```

File: src/py/flwr/server/app.py [hunk 2]
```diff
@@ -556,3 +565,13 @@ def _add_args_fleet_api(parser: argparse.ArgumentParser) -> None:
...
+        help="Fleet API REST SSL certificate file (as a path str). Default:None",
+        default=None,
+    )
+    rest_group.add_argument(
+        "--ssl_keyfile",
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:574]:
```suggestion
        "--ssl-keyfile",
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/py/flwr/client/rest_client/connection.py [hunk 3]
```diff
@@ -80,7 +85,11 @@ def http_request_response(
         """,
     )
 
-    base_url = f"http://{server_address}"
+    base_url = f"https://{server_address}"
+
+    verify: Union[bool, str] = False
```
Comment (tanertopal) [inline src/py/flwr/client/rest_client/connection.py:90]: @charlesbvll is `verify=False` the default of the requests library? I would prefer to stick to the default if that makes sense in our case.

Comment (tanertopal) [reply src/py/flwr/client/rest_client/connection.py:90]: This is assuming the request uses `https`, right?

File: src/py/flwr/server/app.py [hunk 4]
```diff
@@ -375,7 +381,12 @@ def run_server() -> None:
     )
 
     # Block
-    driver_server.wait_for_termination()
+    while True:
+        if bckg_threads:
+            for thread in bckg_threads:
+                if not thread.is_alive():
+                    sys.exit(1)
+        driver_server.wait_for_termination(timeout=1)
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:389]: @tanertopal is this related to HTTPS, or is it a general improvement? If it's not related to HTTPS, let's move it to a separate PR

File: src/py/flwr/server/app.py [hunk 5]
```diff
@@ -500,6 +512,12 @@ def _run_fleet_api_rest(
     # See: https://www.starlette.io/applications/#accessing-the-app-instance
     fast_api_app.state.STATE_FACTORY = state_factory
 
+    host, port_str = address.split(":")
+    port = int(port_str)
+
+    if not isfile(ssl_keyfile) or not isfile(ssl_certfile):
+        raise FileNotFoundError("One of SSL Keyfile or Certfile is not an actual file.")
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:519]: Let's split this into two if statements and tell the user exactly which path does not point to a file

File: src/py/flwr/client/app.py [hunk 6]
```diff
@@ -155,6 +155,11 @@ class `flwr.client.Client`.
                 "To use the REST API you must install the "
                 "extra dependencies by running `pip install flwr['rest']`."
             ) from missing_dep
+        if server_address[:4] != "http":
+            sys.exit(
+                "When using the REST API please provide `https://` or "
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:160]:
```suggestion
                "When using the REST API, please provide `https://` or "
```

File: src/py/flwr/server/app.py [hunk 7]
```diff
@@ -598,3 +618,13 @@ def _add_args_fleet_api(parser: argparse.ArgumentParser) -> None:
...
+    rest_group.add_argument(
+        "--ssl-certfile",
+        help="Fleet API REST SSL certificate file (as a path str). Default:None",
+        default=None,
+    )
+    rest_group.add_argument(
+        "--ssl-keyfile",
+        help="Fleet API REST SSL private key file (as a path str). Default:None",
+        default=None,
+    )
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:630]: Shouldn't we add these as general args and make them the same across gRPC and REST?

File: src/py/flwr/server/app.py (hunk 4 above)
Comment (tanertopal) [reply src/py/flwr/server/app.py]: @danieljanes I removed this.

File: src/py/flwr/server/app.py (hunk 5 above)
Comment (tanertopal) [reply src/py/flwr/server/app.py]: @danieljanes I rewrote it.

File: src/py/flwr/server/app.py (hunk 7 above)
Comment (tanertopal) [reply src/py/flwr/server/app.py:627]: @charlesbvll can you look into it?

Comment (danieljanes) [reply src/py/flwr/server/app.py:629]: @tanertopal @charlesbvll we could even put those new args (and their usage in gRPC) into a new PR

File: src/py/flwr/client/app.py [hunk 8]
```diff
@@ -88,10 +88,10 @@ def start_client(
     server_address: str,
     client: Client,
     grpc_max_message_length: int = GRPC_MAX_MESSAGE_LENGTH,
-    root_certificates: Optional[bytes] = None,
+    root_certificates: Optional[Union[bytes, str]] = None,
     rest: bool = False,
 ) -> None:
-    """Start a Flower Client which connects to a gRPC server.
+    """Start a Flower Client which connects to a Flower server.
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:94]:
```suggestion
    """Start a Flower client node which connects to a Flower server.
```

File: src/py/flwr/client/app.py [hunk 9]
```diff
@@ -109,25 +109,25 @@ class `flwr.client.Client`.
...
+        established to a SSL-enabled Flower server.
     rest : bool (default: False)
         Defines whether or not the client is interacting with the server using the
         experimental REST API. This feature is experimental, it might change
         considerably in future versions of Flower.
 
     Examples
     --------
-    Starting a client with insecure server connection:
+    Starting a gRPC client with insecure server connection:
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:123]:
```suggestion
    Starting a gRPC client with an insecure server connection:
```

File: src/py/flwr/client/app.py [hunk 10]
```diff
@@ -109,25 +109,25 @@ class `flwr.client.Client`.
...
 
     >>> start_client(
     >>>     server_address=localhost:8080,
     >>>     client=FlowerClient(),
     >>> )
 
-    Starting a SSL-enabled client:
+    Starting a SSL-enabled gRPC client:
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:130]:
```suggestion
    Starting an SSL-enabled gRPC client:
```

File: src/py/flwr/client/grpc_client/connection.py [hunk 11]
```diff
@@ -61,9 +62,9 @@ def grpc_connection(
         increased limit and block larger messages.
         (default: 536_870_912, this equals 512MB)
     root_certificates : Optional[bytes] (default: None)
-        The PEM-encoded root certificates as a byte string. If provided, a secure
-        connection using the certificates will be established to a SSL-enabled
-        Flower server.
+        The PEM-encoded root certificates as a byte string or a path string.
+        If provided, a secure connection using the certificates will be
+        established to a SSL-enabled Flower server.
```
Comment (danieljanes) [inline src/py/flwr/client/grpc_client/connection.py:67]:
```suggestion
        established to an SSL-enabled Flower server.
```

File: src/py/flwr/client/app.py [hunk 12]
```diff
@@ -218,9 +223,9 @@ def start_numpy_client(
         same value (see `flwr.server.start_server`), otherwise it will not
         know about the increased limit and block larger messages.
     root_certificates : bytes (default: None)
-        The PEM-encoded root certificates a byte string. If provided, a secure
-        connection using the certificates will be established to a
-        SSL-enabled Flower server.
+        The PEM-encoded root certificates as a byte string or a path string.
+        If provided, a secure connection using the certificates will be
+        established to a SSL-enabled Flower server.
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:228]:
```suggestion
        established to an SSL-enabled Flower server.
```

File: src/py/flwr/client/app.py [hunk 13]
```diff
@@ -109,25 +109,25 @@ class `flwr.client.Client`.
...
         same value (see `flwr.server.start_server`), otherwise it will not
         know about the increased limit and block larger messages.
-    root_certificates : bytes (default: None)
-        The PEM-encoded root certificates as a byte string. If provided, a secure
-        connection using the certificates will be established to a
-        SSL-enabled Flower server.
+    root_certificates : Optional[Union[bytes, str]] (default: None)
+        The PEM-encoded root certificates as a byte string or a path string.
+        If provided, a secure connection using the certificates will be
+        established to a SSL-enabled Flower server.
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:115]:
```suggestion
        established to an SSL-enabled Flower server.
```

File: src/py/flwr/client/rest_client/connection.py [hunk 14]
```diff
@@ -60,12 +62,15 @@ def http_request_response(
...
-        The IPv6 address of the server. If the Flower server runs on the same machine
-        on port 8080, then `server_address` would be `"[::]:8080"`.
+        The IPv6 address of the server with `http://` or `https://`.
+        If the Flower server runs on the same machine
+        on port 8080, then `server_address` would be `"http://[::]:8080"`.
     max_message_length : int
         Ignored, only present to preserve API-compatibility.
-    root_certificates : Optional[bytes] (default: None)
-        Ignored, for now.
+    root_certificates : Optional[Union[bytes, str]] (default: None)
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:70]: Shouldn't the type be `Optional[str]`?

File: src/py/flwr/client/rest_client/connection.py [hunk 15]
```diff
@@ -60,12 +62,15 @@ def http_request_response(
...
+        Path of the root certificate. If provided, a secure
+        connection using the certificates will be established to a SSL-enabled
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:72]:
```suggestion
        connection using the certificates will be established to an SSL-enabled
```

File: src/py/flwr/client/rest_client/connection.py [hunk 16]
```diff
@@ -60,12 +62,15 @@ def http_request_response(
...
+        Flower server. Bytes won't work for the REST API.
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:73]: We can get rid of this if we change the type:
```suggestion
        Flower server.
```

File: src/py/flwr/client/rest_client/connection.py [hunk 17]
```diff
@@ -80,7 +85,21 @@ def http_request_response(
...
+    # https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification
+    verify: Union[bool, str] = True
+    if isinstance(root_certificates, str):
+        verify = root_certificates
+    elif isinstance(root_certificates, bytes):
+        log(
+            ERROR,
+            "For the REST API, the root certificates "
+            "must be provided as a string path to the client.",
+        )
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:102]: Again, it seems like we can avoid handling bytes here altogether.

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/py/flwr/server/app.py [hunk 18]
```diff
@@ -500,6 +507,26 @@ def _run_fleet_api_rest(
     # See: https://www.starlette.io/applications/#accessing-the-app-instance
     fast_api_app.state.STATE_FACTORY = state_factory
 
+    host, port_str = address.split(":")
+    port = int(port_str)
+
+    validation_exceptions = []
+
+    if ssl_keyfile is not None and not isfile(ssl_keyfile):
+        msg = "Parameter `ssl_keyfile` is not a file. The path might be wrong."
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:516]:
```suggestion
        msg = "Argument `ssl-keyfile` is not a file. The path might be wrong."
```

File: src/py/flwr/server/app.py [hunk 19]
```diff
@@ -500,6 +507,26 @@ def _run_fleet_api_rest(
...
+        log(ERROR, msg)
+        validation_exceptions.append(ValueError(msg))
+
+    if ssl_certfile is not None and not isfile(ssl_certfile):
+        msg = "Parameter `ssl_certfile` is not a file. The path might be wrong."
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:521]:
```suggestion
        msg = "The path provided as the `--ssl-certfile` argument does not point to a file."
```

File: src/py/flwr/server/app.py [hunk 20]
```diff
@@ -500,6 +507,26 @@ def _run_fleet_api_rest(
     # See: https://www.starlette.io/applications/#accessing-the-app-instance
     fast_api_app.state.STATE_FACTORY = state_factory
 
+    host, port_str = address.split(":")
+    port = int(port_str)
+
+    validation_exceptions = []
+
```
Comment (danieljanes) [inline src/py/flwr/server/app.py]:
```suggestion
    # Check if ssl-keyfile and ssl-certfile point to actual files 
    validation_exceptions = []
```

File: src/py/flwr/client/rest_client/connection.py (hunk 14 above)
Comment (tanertopal) [reply src/py/flwr/client/rest_client/connection.py:70]: gRPC accepts `bytes`, which was the previous type, and the rest stack expects `str`.

File: src/py/flwr/client/rest_client/connection.py (hunk 16 above)
Comment (tanertopal) [reply src/py/flwr/client/rest_client/connection.py:73]: @danieljanes we could but it would be a breaking change here.

File: src/py/flwr/server/app.py [hunk 21]
```diff
@@ -508,9 +524,36 @@ def _run_fleet_api_rest(
...
     )
 
 
+def _validate_ssl_files(
+    ssl_keyfile: Optional[str], ssl_certfile: Optional[str]
+) -> List[ValueError]:
+    validation_exceptions = []
+
+    if ssl_keyfile is not None and not isfile(ssl_keyfile):
+        msg = "Parameter `ssl_keyfile` is not a file. The path might be wrong."
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:538]:
```suggestion
        msg = "Path argument `--ssl-keyfile` does not point to a file."
```

File: src/py/flwr/server/app.py [hunk 22]
```diff
@@ -508,9 +524,36 @@ def _run_fleet_api_rest(
...
+        log(ERROR, msg)
+        validation_exceptions.append(ValueError(msg))
+
+    if ssl_certfile is not None and not isfile(ssl_certfile):
+        msg = "Parameter `ssl_certfile` is not a file. The path might be wrong."
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:543]:
```suggestion
        msg = "Path argument `--ssl-certfile` does not point to a file."
```

File: src/py/flwr/server/app.py [hunk 23]
```diff
@@ -508,9 +524,36 @@ def _run_fleet_api_rest(
...
+        log(ERROR, msg)
+        validation_exceptions.append(ValueError(msg))
+
+    if not (bool(ssl_keyfile) == bool(ssl_certfile)):
+        msg = (
+            "When setting one of `ssl_keyfile` and `ssl_certfile` both have to be used."
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:549]:
```suggestion
            "When setting one of `--ssl-keyfile` and `--ssl-certfile`, both have to be used."
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/py/flwr/client/rest_client/connection.py (hunk 14 above)
Comment (danieljanes) [reply src/py/flwr/client/rest_client/connection.py:70]: I see, I missed the fact that this function and the gRPC function are used interchangeably, so we have to do the check inside the function.

File: src/py/flwr/server/app.py (hunk 7 above)
Comment (charlesbvll) [reply src/py/flwr/server/app.py:653]: I will open the new PR as soon as this one is merged

**Review by tanertopal — COMMENTED** [review]
@danieljanes lgtm and can be merged from my point of view.
